### ISSUE TITLE
model history note

### ISSUE THREAD

**Issue by karino2** [karino2/tegashiki#1]
Note of each model.

**Comment by karino2** [comment]
### model_vec_mask3

Initial model. Data is not yet reduced by Ramer–Douglas–Peucker algorithm.
Mask is just mean properly ignoring padding of sequence. This is the initial model and basically many parameter (more than training set size, I guess).

![model_vec_mask3](https://user-images.githubusercontent.com/1468643/58378646-42145800-7fd2-11e9-8764-f9f403f3c371.png)

**Comment by karino2** [comment]
### model_vec_small

Reduce embed and hidden layer parameter from model_vec_mask3.
Similar tendency with bad score.

![model_vec_small](https://user-images.githubusercontent.com/1468643/58378659-6bcd7f00-7fd2-11e9-9301-952dac43a8d6.png)

**Comment by karino2** [comment]
### small_embed

Just reduce embed size from model_vec_mask3. Similar tendency to model_vec_mask3. A little bad score.
Accuracy score is very fragile.

![small_embed](https://user-images.githubusercontent.com/1468643/58378665-8f90c500-7fd2-11e9-87a2-734a2c99caf1.png)

**Comment by karino2** [comment]
### rdp_mask3

Reduce data by Ramer–Douglas–Peucker algorithm, epsilon 0.05.
Model is the same as model_vec_mask3.

From here, we use this dataset.

This is the best acc of 0.39, but it seems too random picking.

![rdp_mask3](https://user-images.githubusercontent.com/1468643/58378907-ec8e7a00-7fd6-11e9-9553-7e20b510e1df.png)

**Comment by karino2** [comment]
### rdp_small

small parameter trial for rdp reductioned dataset.
More stable, but bad score.

![rdp_small](https://user-images.githubusercontent.com/1468643/58379005-02506f00-7fd8-11e9-969b-54015dccdd72.png)

**Comment by karino2** [comment]
### rdp_small_embed

Just reduce embed size from rdp_mask3.
Accuracy is fragile, but score is similar (a little worse than mask3).

![small_embed](https://user-images.githubusercontent.com/1468643/58379016-2dd35980-7fd8-11e9-8b1e-2221c6aceaa4.png)

**Comment by karino2** [comment]
### tcn_avg

TCN for encoder. with avg pooling. depth=6. acc 0.335.

![tcn_avg](https://user-images.githubusercontent.com/1468643/58400106-24f48d80-8095-11e9-97ce-a830522c2241.png)

**Comment by karino2** [comment]
### tcn_dropout_lr0001

Add dropout in TCN (not in decoder, though). Raise learning rate 0.001 instead of 0.00009.
learning rate seems too high.

![tcn_dropout_lr0001](https://user-images.githubusercontent.com/1468643/58400245-78ff7200-8095-11e9-9000-61261f1c42a4.png)

**Comment by karino2** [comment]
### tcn_dropout

Lower learning rate from above trial.

Dropout seems learning a little regularize, but the final score is not improved.

![tcn_dropout](https://user-images.githubusercontent.com/1468643/58400513-325e4780-8096-11e9-9dc8-fb02dafbcd88.png)

**Comment by karino2** [comment]
### cnnrnn2_dropout

Add dropout to rdp_mask3. We call previous conv1d+RNN encoder model as cnnrnn2 from now on.

acc 0.3228. Accuracy goes down with dropout. I don't know the reason, but the score of this model is fragile and it might just pick randomly good value in previous trial.

![cnnrnn2_dropout](https://user-images.githubusercontent.com/1468643/58446151-83277c00-813a-11e9-8fc8-428399e03f22.png)

**Comment by karino2** [comment]
### tcn_drop2_initstate

From tcn_dropout, add dropout in decoder side, and apply FC in last output of tcn and plug into decoder's init state.
acc 0.344

Training becomes stable, but score is the same.

![tcn_drop2_initstate_fix](https://user-images.githubusercontent.com/1468643/58448094-b7eb0180-8141-11e9-933b-cb142b914a72.png)

**Comment by karino2** [comment]
### tcn_drop2_d8

Increase dilated layer to 8 to cover wider range for encoder init state.
acc 0.347. A little better, but almost the same.

![tcn_drop2_d8](https://user-images.githubusercontent.com/1468643/58448439-ec12f200-8142-11e9-9287-a23253904528.png)

**Comment by karino2** [comment]
### tcn_regu

Add l2 regularization of 0.1 to weight, kernel, activation with previous model.
acc = 0.34. The same score.

![tcn_regu](https://user-images.githubusercontent.com/1468643/58452723-5338a280-8153-11e9-9bff-581025579fab.png)

**Comment by karino2** [comment]
## Parser fix

I realize that parser for tfrecord had bug and stroke is wrongly concatenated.

I had tried to reshape to

```
[[x1, y1, type1], [x2, y2, type2], ..., [xn, yn, typen]]
```

But the reality was

```
[[x1, x2, x3], [x4, x5, x6], ..., [type(n-2), type(n-1), typen]]
```

So it might be hard to match x, y, type relations.
All previous experiments was conducted with this bug.

**Comment by karino2** [comment]
### tcn_regu_fixparser

I fix parser with above model.
acc=0.335. Almost the same.

Learning curve is changed, so fix seems applied.
Hard to believe that the final score is almost the same.

![tcn_regu_fixparser](https://user-images.githubusercontent.com/1468643/58453023-7879e080-8154-11e9-888a-f59741ac3383.png)

**Comment by karino2** [comment]
### cnnrnn_regular_fixparser

cnnrnn2 with weight regularization the same as above model. And fix parser.
acc = 0.407.

It seems this is the best model and this model can take stroke data into account.
Almost perfectly fit to training set, though it's severely overfitted. (final training loss is 0.2179).

![cnnrnn2_regu_fixparser](https://user-images.githubusercontent.com/1468643/58453836-7feeb900-8157-11e9-9638-73203dc907b7.png)

**Comment by karino2** [comment]
### cnnrnn2_regular_fixparser_small

Reduce parameter size with above model.
acc=0.379

Less overfit, and score is stable, but getting worse.

![cnnrnn2_fixparser_small](https://user-images.githubusercontent.com/1468643/58531350-ee497f00-821c-11e9-80b5-779fc1494e8e.png)

**Comment by karino2** [comment]
### cnnrnn2_drop05

Increase dropout rate from 0.1 to 0.5 in above model.
acc=0.377

Overfit is not improved.

![cnnrnn2_small_drop05](https://user-images.githubusercontent.com/1468643/58531471-6021c880-821d-11e9-9a35-ffa6e3e331e9.png)

**Comment by karino2** [comment]
### cnnrnn2_drop09

Increase dropout rate to 0.9 with above model (cnnrnn2-small).
acc=0.355.
Becomes learned more gradually, but converge to the same score.
Overfitting is not fixed with dropout.

![cnnrnn2_small_drop09](https://user-images.githubusercontent.com/1468643/58531561-b7279d80-821d-11e9-9d17-541992c87a88.png)

**Comment by karino2** [comment]
This cell is move to #3

**Comment by karino2** [comment]
## Use feature extractor

From discussion, I guess it might be better to compress each stroke to more compressed data.
So create feature extractor from subtask one symbol dataset.

https://github.com/karino2/tegashiki/blob/master/tegashiki_symbol.ipynb

Basic design note:

- Conv-MaxPool-Conv-MaxPool-Conv-GlobalMaxPooling1D network to extract feature as 256 dim vector
- feature extractor is trained by connected to GRU, one symbol predict task (by last GRU output).  accuracy is 0.758
- Score seems a ilttle bit low. But original dataset is not so clean and even I can't understand some symbols.

**Comment by karino2** [comment]
### padstroke_small_rnn_small_dropout05

- Use feature extractor
- Create stroke data the same as feature extractor task (padding and make one stroke same size (50))
- Add GRU encoder and decoder after extracted feature, then add attention (stroke num is small enough that we can train without pooling)
- Parameter is almost the same as cnnrnn2_regular_fixparser_small

acc: 0.36

No Improvement.

![padstroke_small_rnn_small_dropout05](https://user-images.githubusercontent.com/1468643/59813342-0ad46500-934c-11e9-88c4-819486185b75.png)

**Comment by karino2** [comment]
### samelen_small_rnn_small_dropout05

I couldn't believe previous result. It must be bug!

Previous dataset is different stroke num for training set and validation set.
I guess this might cause some loss function calculation mismatch (I wonder attention code might not handle stroke mask correctly).

acc 0.34

No improvement. Seems not a bug.

![samelen_small_rnn_small_dropout05](https://user-images.githubusercontent.com/1468643/59813499-ae257a00-934c-11e9-96ff-f4429c234ad7.png)

**Comment by karino2** [comment]
No matter what I try, the accuracy tend to converge to 0.3.
And training error lower much further, but not validation loss improvement.

Why?
One guess is that decoder combination is too complex and not so much data, so our model can't learn structure with this model.
The best he can do might be just predict prior distribution of math expression tokens.

To confirm my hypothesis, I create model that just ignore input stroke.

### samelen_nostroke

Just ignore input stroke and only train from decoder input-output.
This is some form of math exp language model.
It is most idiot model.

acc 0.38

![samelen_nostroke](https://user-images.githubusercontent.com/1468643/59813724-97cbee00-934d-11e9-8843-c141218a9c44.png)

Getting better!

All attempt so far seems just not working, ignore stroke input and just predict math expression general symbol distribution!

**Comment by karino2** [comment]
### varlen_10_small_rnn_small_dropout05

Previous experiment suggest current dataset might too complex.
So filter dataset that tex symbol len seems too large, and just keep less than 10.
Dataset becomes only 3000. It seems this is too small to learn. But I try anyway.

acc 0.31

![varlen_10_small_rnn_small_dropout05](https://user-images.githubusercontent.com/1468643/59814027-bb436880-934e-11e9-9dbd-a425adb3af3b.png)

No Improvement. It seems this task is impossible to solve!

**Comment by karino2** [comment]
Go to #3
